File: backend/app/models/guests.py
"""Clients : particuliers, societes, pieces d'identite."""

from __future__ import annotations

import datetime as dt
import uuid

from sqlalchemy import (
    BigInteger,
    Boolean,
    Date,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import HotelScoped, SyncBase
from app.models.enums import IdDocumentType, UploadState
from app.models.rooms import _enum


class Company(SyncBase, HotelScoped):
    """Societe, agence de voyage ou tour-operateur.

    Absente du cahier des charges, mais indissociable d'une facturation
    hoteliere reelle : le debiteur d'une facture n'est pas toujours l'occupant
    de la chambre. Sans cette table, impossible d'emettre une facture au nom
    d'une entreprise ni de gerer un encours client.
    """

    __tablename__ = "companies"
    __table_args__ = (UniqueConstraint("hotel_id", "code"),)

    code: Mapped[str] = mapped_column(String(32))
    name: Mapped[str] = mapped_column(String(160), index=True)
    tax_id: Mapped[str | None] = mapped_column(String(40), default=None)
    address: Mapped[str | None] = mapped_column(String(255), default=None)
    city: Mapped[str | None] = mapped_column(String(80), default=None)
    country: Mapped[str | None] = mapped_column(String(80), default=None)
    contact_name: Mapped[str | None] = mapped_column(String(120), default=None)
    phone: Mapped[str | None] = mapped_column(String(40), default=None)
    email: Mapped[str | None] = mapped_column(String(160), default=None)
    credit_limit: Mapped[int] = mapped_column(BigInteger, default=0)
    payment_terms_days: Mapped[int] = mapped_column(default=0)
    discount_rate: Mapped[int] = mapped_column(Integer, default=0)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    notes: Mapped[str | None] = mapped_column(Text, default=None)


class Guest(SyncBase, HotelScoped):
    """Fiche client (F1.6).

    C'est la seule table porteuse de donnees personnelles en clair. Ce
    cloisonnement est deliberé : il concentre sur un perimetre reduit le
    chiffrement au repos (exigence 6.2) et le droit a l'effacement (RGPD,
    paragraphe 8), au lieu de les disperser dans toute la base.
    """

    __tablename__ = "guests"
    __table_args__ = (
        UniqueConstraint("hotel_id", "code"),
        Index("ix_guests_last_name", "last_name"),
        Index("ix_guests_phone", "phone"),
        Index("ix_guests_email", "email"),
        Index("ix_guests_id_document_number", "id_document_number"),
    )

    code: Mapped[str] = mapped_column(String(32))
    title: Mapped[str | None] = mapped_column(String(16), default=None)
    first_name: Mapped[str] = mapped_column(String(80))
    last_name: Mapped[str] = mapped_column(String(80))
    birth_date: Mapped[dt.date | None] = mapped_column(Date, default=None)
    birth_place: Mapped[str | None] = mapped_column(String(120), default=None)
    nationality: Mapped[str | None] = mapped_column(String(80), default=None)
    gender: Mapped[str | None] = mapped_column(String(16), default=None)

    id_document_type: Mapped[IdDocumentType | None] = mapped_column(
        _enum(IdDocumentType, "id_document_type"), default=None
    )
    id_document_number: Mapped[str | None] = mapped_column(String(64), default=None)
    id_document_expiry: Mapped[dt.date | None] = mapped_column(Date, default=None)

    email: Mapped[str | None] = mapped_column(String(160), default=None)
    phone: Mapped[str | None] = mapped_column(String(40), default=None)
    phone_alt: Mapped[str | None] = mapped_column(String(40), default=None)
    address: Mapped[str | None] = mapped_column(String(255), default=None)
    city: Mapped[str | None] = mapped_column(String(80), default=None)
    postal_code: Mapped[str | None] = mapped_column(String(20), default=None)
    country: Mapped[str | None] = mapped_column(String(80), default=None)

    company_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("companies.id", ondelete="SET NULL"), default=None, index=True
    )

    # Preferences libres : etage eleve, chambre non-fumeur, oreiller ferme,
    # allergies. Champ JSON plutot qu'un schema fige, parce que chaque hotel
    # suit ses propres criteres et qu'ils changent souvent.
    preferences: Mapped[dict | None] = mapped_column(JSONB, default=None)
    notes: Mapped[str | None] = mapped_column(Text, default=None)
    is_vip: Mapped[bool] = mapped_column(Boolean, default=False)
    is_blacklisted: Mapped[bool] = mapped_column(Boolean, default=False)
    blacklist_reason: Mapped[str | None] = mapped_column(String(255), default=None)

    # RGPD : consentement explicite et horodate, separe de l'adresse courriel.
    marketing_consent: Mapped[bool] = mapped_column(Boolean, default=False)
    consent_at: Mapped[dt.datetime | None] = mapped_column(default=None)
    anonymized_at: Mapped[dt.datetime | None] = mapped_column(default=None)

    company: Mapped[Company | None] = relationship(lazy="joined")

    @property
    def full_name(self) -> str:
        return " ".join(p for p in (self.title, self.first_name, self.last_name) if p)


class GuestDocument(SyncBase):
    """Scan ou photo d'une piece d'identite.

    Comme toutes les pieces jointes du systeme, le fichier existe d'abord
    localement sur la tablette (`file_path_local`) puis recoit une URL serveur
    apres televersement. Un binaire ne transite pas par la file d'attente JSON
    de synchronisation : c'est un worker distinct qui le remonte, et
    `upload_state` suit son avancement.
    """

    __tablename__ = "guest_documents"

    guest_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("guests.id", ondelete="CASCADE"), index=True
    )
    doc_type: Mapped[IdDocumentType] = mapped_column(_enum(IdDocumentType, "doc_type"))
    file_path_local: Mapped[str | None] = mapped_column(String(255), default=None)
    file_url: Mapped[str | None] = mapped_column(String(512), default=None)
    mime_type: Mapped[str | None] = mapped_column(String(80), default=None)
    size_bytes: Mapped[int | None] = mapped_column(default=None)
    upload_state: Mapped[UploadState] = mapped_column(
        _enum(UploadState, "upload_state"), default=UploadState.PENDING
    )
    captured_at: Mapped[dt.datetime | None] = mapped_column(default=None)

File: frontend/lib/data/local/tables/guests.dart
/// Clients : particuliers, societes, pieces d'identite.
library;

import 'package:drift/drift.dart';

import '../columns.dart';
import '../enums.dart';

/// Societe, agence de voyage ou tour-operateur.
///
/// Absente du cahier des charges, mais indissociable d'une facturation
/// hoteliere reelle : le debiteur d'une facture n'est pas toujours l'occupant
/// de la chambre. Sans cette table, impossible d'emettre une facture au nom
/// d'une entreprise ni de suivre un encours.
@DataClassName('CompanyRow')
class Companies extends Table
    with SyncedTableColumns, HotelScoped, RefTableColumns {
  TextColumn get code => text().withLength(max: 32)();
  TextColumn get name => text().withLength(max: 160)();
  TextColumn get taxId => text().withLength(max: 40).nullable()();
  TextColumn get address => text().withLength(max: 255).nullable()();
  TextColumn get city => text().withLength(max: 80).nullable()();
  TextColumn get country => text().withLength(max: 80).nullable()();
  TextColumn get contactName => text().withLength(max: 120).nullable()();
  TextColumn get phone => text().withLength(max: 40).nullable()();
  TextColumn get email => text().withLength(max: 160).nullable()();

  /// Encours autorise, en francs CFA, entiers.
  IntColumn get creditLimit => integer().withDefault(const Constant(0))();
  IntColumn get paymentTermsDays => integer().withDefault(const Constant(0))();

  /// Remise negociee, en points de base : 5 % vaut 500.
  IntColumn get discountRate => integer().withDefault(const Constant(0))();
  TextColumn get notes => text().nullable()();
}

/// Fiche client (F1.6).
///
/// C'est la seule table de la base qui porte des donnees personnelles en
/// clair. Ce cloisonnement est voulu : il concentre sur un perimetre etroit le
/// chiffrement au repos (exigence 6.2) et le droit a l'effacement (RGPD,
/// paragraphe 8). Disperser nom et telephone dans dix tables rendrait les deux
/// impossibles a garantir.
@DataClassName('GuestRow')
class Guests extends Table with SyncedTableColumns, HotelScoped {
  TextColumn get code => text().withLength(max: 32)();
  TextColumn get title => text().withLength(max: 16).nullable()();
  TextColumn get firstName => text().withLength(max: 80)();
  TextColumn get lastName => text().withLength(max: 80)();
  TextColumn get birthDate => text().withLength(max: 10).nullable()();
  TextColumn get birthPlace => text().withLength(max: 120).nullable()();
  TextColumn get nationality => text().withLength(max: 80).nullable()();
  TextColumn get gender => text().withLength(max: 16).nullable()();

  TextColumn get idDocumentType => textEnum<IdDocumentType>().nullable()();
  TextColumn get idDocumentNumber => text().withLength(max: 64).nullable()();
  TextColumn get idDocumentExpiry => text().withLength(max: 10).nullable()();

  TextColumn get email => text().withLength(max: 160).nullable()();
  TextColumn get phone => text().withLength(max: 40).nullable()();
  TextColumn get phoneAlt => text().withLength(max: 40).nullable()();
  TextColumn get address => text().withLength(max: 255).nullable()();
  TextColumn get city => text().withLength(max: 80).nullable()();
  TextColumn get postalCode => text().withLength(max: 20).nullable()();
  TextColumn get country => text().withLength(max: 80).nullable()();

  TextColumn get companyId => text().nullable()();

  /// Preferences libres, en JSON : etage eleve, chambre non-fumeur, oreiller
  /// ferme, allergies. Chaque hotel suit ses propres criteres et les fait
  /// evoluer souvent -- un schema fige vieillirait mal.
  TextColumn get preferences => text().nullable()();
  TextColumn get notes => text().nullable()();

  BoolColumn get isVip => boolean().withDefault(const Constant(false))();
  BoolColumn get isBlacklisted =>
      boolean().withDefault(const Constant(false))();
  TextColumn get blacklistReason => text().withLength(max: 255).nullable()();

  /// Consentement commercial, horodate et distinct de la simple presence
  /// d'une adresse courriel (RGPD).
  BoolColumn get marketingConsent =>
      boolean().withDefault(const Constant(false))();
  DateTimeColumn get consentAt => dateTime().nullable()();
  DateTimeColumn get anonymizedAt => dateTime().nullable()();
}

/// Scan ou photo d'une piece d'identite.
///
/// Le fichier existe d'abord sur la tablette, puis recoit une URL serveur
/// apres televersement. Un binaire ne transite pas par la file d'attente JSON
/// de synchronisation : c'est un traitement distinct, dont `uploadState` suit
/// l'avancement.
@DataClassName('GuestDocumentRow')
class GuestDocuments extends Table with SyncedTableColumns {
  TextColumn get guestId =>
      text().references(Guests, #id, onDelete: KeyAction.cascade)();
  TextColumn get docType => textEnum<IdDocumentType>()();
  TextColumn get filePathLocal => text().withLength(max: 255).nullable()();
  TextColumn get fileUrl => text().withLength(max: 512).nullable()();
  TextColumn get mimeType => text().withLength(max: 80).nullable()();
  IntColumn get sizeBytes => integer().nullable()();
  TextColumn get uploadState =>
      textEnum<UploadState>().withDefault(const Constant('PENDING'))();
  DateTimeColumn get capturedAt => dateTime().nullable()();
}
